src/Forms/GridField/GridField.php

```diff
@@ -444,6 +451,66 @@ private function initState(): void
         }
     }
 
+    /**
+     * Adds state for this gridfield from the request variables.
+     *
+     * If there is state already set on this GridField, that takes precedent
+     * over state from the request.
+     */
+    private function addStateFromRequest(): void
+    {
+        $request = $this->getRequest();
+        if (($request instanceof NullHTTPRequest) && Controller::has_curr()) {
+            $request = Controller::curr()->getRequest();
+        }
+        
+        $stateStr = $this->getStateManager()->getStateFromRequest($this, $request);
+
+        if ($stateStr) {
+            $oldState = $this->getState(false);
+            // Create a dummy state so that we can merge the current state with the request state.
+            $newState = new GridState($this, $stateStr);
+            // Put the current state on top of the request state.
+            $newState->setValue($oldState->Value());
+            $this->state = $newState;
+        }
+    }
+
+    /**
+     * Add get and post parameters if exist to the URL.
+     * URL keeps state for GridField, that exist in few different tabs.
+     */
+    public function addRequestToURL(string $link): string
+    {
+        $request = $this->getRequest();
+
+        if (($request instanceof NullHTTPRequest) && Controller::has_curr()) {
+            $request = Controller::curr()->getRequest();
+        }
+
+        $postparams = [];
+
+        if ($request->postVars()) {
+            foreach ($request->postVars() as $key => $val) {
+                if (stripos($key, 'gridState') === 0) {
+                    $postparams[$key] = $val;
+                }
+            }
+        }
+
+        $params = array_merge($request->getVars(), $postparams);
+
+        if ($params && is_array($params) && count($params) > 0) {
+            foreach ($params as $key => $val) {
+                $link = HTTP::setGetVar($key, $val, $link);
+            }
+        }
```

Comment:
You can simplify this by using `$request->requestVars()` which already combines `GET` and `POST` vars for you.

Reply:
I was thinking about `requestVars() `method, but...
I don't want to go through the whole `GET` request, because I know that I need everything from it. And also I know that I need only values with keys starting with `'gridState-.` from `POST` params. Also I know that `POST` request should have the newest params and I know that  `array_merge($request->getVars(), $postparams)` replaces duplicates from GET request with new POST request.
But, probably, my assumptions are wrong.

Reply:
I think it's worth testing with it - if it works correctly with `requestVars()` then I'd recommend using that. If it _doesn't_ work with `requestVars()` we probably should have some comments in the code explaining what's going on.

Reply:
DONE